blackjax/mcmc/metrics.py

```diff
@@ -208,6 +213,120 @@ def is_turning(
+    is_turning
+        A function that determines whether a trajectory is turning back on
+        itself given the values of the momentum along the trajectory.
+
+    """
+    (ndim,) = jnp.shape(diagonal_scale_std)
+    (ndim_, n_eigs) = jnp.shape(eigenvectors)
+    if ndim != ndim_:
+        raise ValueError("Shape mismatch in metric.")
+
+    (n_eigs_,) = jnp.shape(eigenvalues)
+    if n_eigs != n_eigs_:
+        raise ValueError("Shape mismatch in metric.")
+
+    # Compute (V(\Sigma - I)V^T + I)x
+    def inner_matrix_mult(vals, vecs, x):
+        projected = x @ vecs
+        scaled = (vals - 1) * projected
+        projected_back = vecs @ scaled
+        return projected_back + x
+
+    def inv_mass_matrix_mult(x):
+        scaled = x * diagonal_scale_std
+        product = inner_matrix_mult(eigenvalues, eigenvectors, scaled)
+        return product * diagonal_scale_std
+
+    def momentum_generator(rng_key: PRNGKey, position: ArrayLikeTree) -> ArrayTree:
+        unit_draws = generate_gaussian_noise(rng_key, position)
+        sqrt_vals = jnp.sqrt(jnp.reciprocal(eigenvalues))
+        sqrt_inv_diag = jnp.sqrt(jnp.reciprocal(diagonal_scale_std))
+        return inner_matrix_mult(sqrt_vals, eigenvectors, unit_draws) * sqrt_inv_diag
+
+    def kinetic_energy(
+        momentum: ArrayLikeTree, position: Optional[ArrayLikeTree] = None
+    ) -> float:
+        del position
+        momentum, _ = ravel_pytree(momentum)
+        velocity = inv_mass_matrix_mult(momentum)
+        kinetic_energy_val = 0.5 * jnp.dot(velocity, momentum)
+        return kinetic_energy_val
+
+    def is_turning(
+        momentum_left: ArrayLikeTree,
+        momentum_right: ArrayLikeTree,
+        momentum_sum: ArrayLikeTree,
+        position_left: Optional[ArrayLikeTree] = None,
+        position_right: Optional[ArrayLikeTree] = None,
+    ) -> bool:
+        """Generalized U-turn criterion :cite:p:`betancourt2013generalizing,nuts_uturn`.
+
+        Parameters
+        ----------
+        momentum_left
+            Momentum of the leftmost point of the trajectory.
+        momentum_right
+            Momentum of the rightmost point of the trajectory.
+        momentum_sum
+            Sum of the momenta along the trajectory.
+
+        """
+        del position_left, position_right
+
+        m_left, _ = ravel_pytree(momentum_left)
+        m_right, _ = ravel_pytree(momentum_right)
+        m_sum, _ = ravel_pytree(momentum_sum)
+
+        velocity_left = inv_mass_matrix_mult(m_left)
+        velocity_right = inv_mass_matrix_mult(m_right)
+
+        # rho = m_sum
+        rho = m_sum - (m_right + m_left) / 2
+        turning_at_left = jnp.dot(velocity_left, rho) <= 0
+        turning_at_right = jnp.dot(velocity_right, rho) <= 0
+        return turning_at_left | turning_at_right
+
+    return Metric(
+        momentum_generator,
+        kinetic_energy,
+        is_turning,
+        data=(diagonal_scale_std, eigenvalues, eigenvectors),
```

Comment:
I guess you want to store these as properties so it is easier for tuning later on? In any case I suggest removing it here, and add it in the subsequent PR when tuning is introduced (so we can discuss whether it is necessary)

Reply:
Yes, that was mostly for debugging and I forgot to take it out. Sorry :-)